### DISCUSSION TITLE
New features for easier object mapping

### DISCUSSION THREAD

**Discussion by Illusive7Man** [dotnet/csharplang#6082] [Language Ideas]
### Current options
When mapping one object to another, e.g. EF Code-First Model object to an object used as an action result, current options are:
- **Manual** Define class with all the properties target object requires, and do the mapping from the source in the constructor
- **AutoMapper** Use a library for runtime mapping where users can define rules manually
- **SourceGenerator** Create a source generator that will add a source that maps object using property assignments

Manual way creates work repetition, runtime libraries introduce new syntax and can hide code usages of properties, and code generation  has its own complexity issues (whether it emits source or not).

### Proposal
A simpler way, which would be my preferred option, for most use cases of mappings would be based on two language features I found in TypeScript. 
- First one is parameter properties, which allows developers to define class properties by using constructor parameters. Those parameters with access declaration are considered to be properties of that class.
- Second is object destructuring, where parameters can be assigned from objects using patterns (like C# tuple deconstructing). Object destructuring assigns values to parameters / variables by relating them to properties of the source object.

With features like these, relationship between source and target would be firmly established because there would be only **one place for selecting which properties to use and what values they should get**. I feel that the significance of these properties would then be derived from the mapping (i.e. the source), and it would not be the same as of those that originate in the target class.

_Code example in TypeScript, along with a discussion on this proposal for TypeScript can be found https://github.com/Microsoft/TypeScript/issues/5326._

### Syntax

I will try to improve it but here's what I got so far based on pattern matching:

```csharp

/*========================== Entities ==========================*/

public class Person
{
    [Key] public int Id { get; set; }
    public string Name { get; set; } = null!;
    public List<Movie> Directed { get; set; } = null!;
    public List<Movie> ActedIn { get; set; } = null!;
}

public class Movie
{
    [Key] public int Id { get; set; }
    public string Title { get; set; } = null!;
    public Person DirectedBy { get; set; } = null!;
    public List<Person> Actors { get; set; } = null!;
}

/*========================== Result types ==========================*/

public class MovieReview
{
    public int Id { get; set; }
    public string ReviewText { get; set; }
    public string ReviewerName { get; set; }

    public MovieReview({
        Movie: {
            Id: public int MovieId,
            Title: public string Title,
            DirectedBy: {Id: public int Id, Name: public string Name} public DirectedBy,
            Actors: {Id: public int Id, Name: public string Name}[] public Actors
        }
    }) {}

    // Alternative constructor with nominal destructing
    public MovieReview({
        Movie: {
            Id: public int MovieId,
            public Title,
            {public Id, public Name} public DirectedBy,
            {public Id, public Name}[] public Actors
        }
    }) {}
}

```

**Comment by HaloFour** [comment]
You get the first one already with records, which have primary constructors which are automatically promoted to properties:

```cs
record Person(string Name, int Age);

var person = new Person("John", 28);
Console.WriteLine($"Hello {person.Name}!");
```

For destructuring, see the following proposals:

https://github.com/dotnet/csharplang/discussions/6055
https://github.com/dotnet/csharplang/discussions/3546
https://github.com/dotnet/csharplang/discussions/153

**Reply by HaloFour** [reply]
If you're looking for splatting, there's also:

https://github.com/dotnet/csharplang/discussions/424
https://github.com/dotnet/csharplang/discussions/1654

I don't know if I've seen any "nominal splatting" proposals, although those might be a little wonky given that the C# naming conventions would typically cause members and parameters to differ by case.

**Reply by Illusive7Man** [reply]
Hey,
You are right about parameter properties being implemented in records.
I could've guessed argument destructuring is being discussed already, but I am interested in having all these features (not splatting) so object mapping is easier. How are you handling object mapping, if it's even that big of a deal to you?

**Reply by Illusive7Man** [reply]
I have added the syntax example, if you want to see how it'd look like.
